### PR TITLE
Add middleware documentation

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

There is currently no description on where and how a customer can apply `Layer`.

## Description

[Rendered](https://github.com/awslabs/smithy-rs/blob/harryb/middleware-docs/design/src/server/middleware.md)

Add a `middleware.md` document.

## Notes

This describes the new service builder API and therefore will remain hidden until its official release.

### CODE REVIEW

File: design/src/server/middleware.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+## Introduction to Tower
+
+Smithy Rust is built on top of [`tower`](https://github.com/tower-rs/tower).
+
+> Tower is a library of modular and reusable components for building robust networking clients and servers.
+
+The `tower` library is centered around two main interfaces, the [`Service`](https://docs.rs/tower/latest/tower/trait.Service.html) trait and the [`Layer`](https://docs.rs/tower/latest/tower/trait.Layer.html) trait.
+
+The `Service` trait can be thought of as an asynchronous function from a request to a response coupled with a mechanism to handle back pressure, while the `Layer` trait can be thought of as a way of way of decorating a `Service`, transforming either the request or response.
```
Comment (crisidev) [inline design/src/server/middleware.md:48]: Adding the fn definition that towers gives in their doc combines nicely with your description here 

```rust
async fn(Request) -> Result<Response, Error>
```

File: design/src/server/middleware.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+The `NewService` modifies the behavior of the inner `Service` `S` while the `NewLayer` takes auxiliary data and constructs `NewService<S>` from `S`.
+
+Customers are then able to stack middleware by composing `Layer`s using combinators such as [`ServiceBuilder::layer`](https://docs.rs/tower/latest/tower/struct.ServiceBuilder.html#method.layer) and [`Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html).
+
+<!-- TODO(Update documentation): There's a `Layer` implementation on tuples about to be merged, give it as an example here. -->
+
+## Applying Middleware
+
+One of the primary goals of Smithy Rust is to provide configurability and extensibility through the application of middleware. The customer is able to apply `Layer`s in a variety of key places during the request/response lifecycle. The following schematic labels each configurable middleware position from A to D:
```
Comment (crisidev) [inline design/src/server/middleware.md:86]:
```suggestion
One of the primary goals is to provide configurability and extensibility through the application of middleware. The customer is able to apply `Layer`s in a variety of key places during the request/response lifecycle. The following schematic labels each configurable middleware position from A to D:
```

File: design/src/server/middleware.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        }
+    }
+    C2 --> [*]: HTTP Response
+    C3 --> [*]: HTTP Response
+    C4 --> [*]: HTTP Response
+```
+
+where `UpgradeLayer` is the `Layer` converting Smithy model structures to HTTP structures and the `RoutingService` is responsible for routing requests to the appropriate operation.
+
+### A) Outer Middleware
```
Comment (crisidev) [inline design/src/server/middleware.md:132]: While I understand it, I believe the difference between A) and B) is not clear in the doc.

Comment (hlbarber) [reply design/src/server/middleware.md:132]: Hows this?
https://github.com/awslabs/smithy-rs/pull/1844/commits/baf1ca4772bb2837f2c3929f613c27cb8b6beebf

File: design/src/server/middleware.md (hunk 1 above)
Comment (unexge) [inline design/src/server/middleware.md:48]:
```suggestion
The `Service` trait can be thought of as an asynchronous function from a request to a response coupled with a mechanism to handle back pressure, while the `Layer` trait can be thought of as a way of decorating a `Service`, transforming either the request or response.
```

Also it'd be nice to link `handle back pressure` to: https://docs.rs/tower/latest/tower/trait.Service.html#backpressure

**Pull request merged**